### ISSUE TITLE
[FR] Support more providers in MLflow deployments server

### ISSUE THREAD

**Issue by FotiosBistas** [mlflow/mlflow#11530]
### Willingness to contribute

Yes. I would be willing to contribute this feature with guidance from the MLflow community.

### Proposal Summary

# Providers

Currently ``MLflow Deployments Server`` supports the following providers: 
- openai
- mosaicml
- anthropic
- cohere
- palm
- azure/ azuread
- mlflow-model-serving
- huggingface-text-generation-inference
- ai21labs
- bedrock
- mistral

A lot of new providers have emerged that might be beneficial to include, maybe we can outline the ones here.

### Motivation

> #### What is the use case for this feature?
The primary use case for this feature is to enable users of MLflow to deploy, manage, and serve their LLM endpoints. This will allow users to leverage different LLM platforms directly within the MLflow ecosystem, choosing the one that best fits their specific requirements in terms of model performance, cost-efficiency, geographic availability, or other criteria.

> #### Why is this use case valuable to support for MLflow users in general?

Supporting a broader range of LLM providers is valuable for MLflow users because it enhances the tool's flexibility and applicability across various scenarios. Users benefit from being able to compare and choose between different LLM providers without leaving the MLflow environment, ensuring that they can select the most suitable provider for their particular needs. 

> #### Why is this use case valuable to support for your project(s) or organization?

Our organization would definitely like to see providers like ``TogetherAI`` or ``mixedbread.ai`` deployed in the server. 

> #### Why is it currently difficult to achieve this use case?

### Details

The implementation appears to be straight forward: 

1. Inside ``mlflow/gateway/providers/`` create a ``<provider_name>.py`` that has: 

```python
class <provider_name>Provider(BaseProvider): 
```

If the provider does not respond with the ``OpenAI`` format: 

```python
class <provider_name>Adapter:
```

Which is then called inside ``<provider_name>Provider`` to enforce proper response structure. 

2. Inside ``mlflow/gateway/config.py`` include: 
```python
class Provider(str, Enum): 
  <provider_name> = "<provider_name>"
  
  ...
  
class <provider_name>Config(ConfigModel):
  <provider_name>_api_key: str
  
  
  ...
  
config_types = {
  ...
  <provider_name>
}  

...

class Model(ConfigModel):
    name: Optional[str] = None
    provider: Union[str, Provider]
    config: Optional[
    ...
    <provider_name> 
    ]
```

3. Inside ``mlflow/providers/__init__.py``: 

```python
from typing import Type

from mlflow.exceptions import MlflowException
from mlflow.gateway.config import Provider
from mlflow.gateway.providers.base import BaseProvider


def get_provider(provider: Provider) -> Type[BaseProvider]:
    ...
    from mlflow.gateway.providers.<provider_name> import <provider_name>Provider

    provider_to_class = {
        ...
        Provider.<provider_name>: <provider_name>Provider,
    }
    if prov := provider_to_class.get(provider):
        return prov

    raise MlflowException.invalid_parameter_value(f"Provider {provider} not found")
    ```
    
    
    Let me know if I have missed anything! 

### What component(s) does this bug affect?

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/deployments`: MLflow Deployments client APIs, server, and third-party Deployments integrations
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

### What interface(s) does this bug affect?

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

### What language(s) does this bug affect?

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

### What integration(s) does this bug affect?

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

**Comment by WeichenXu123** [comment]
Thanks! I need to discuss with our team folks. I think one issue is how to maintain more and more providers. Can we add a plugin interface so that mlflow can load custom provider? e.g. we can modify https://github.com/mlflow/mlflow/blob/83b6c5fe826a741300855b640b705c33294c7b44/mlflow/gateway/providers/__init__.py#L8 to allow it load a custom provider

**Comment by WeichenXu123** [comment]
@FotiosBistas 

I talked with our team folks, supporting custom provider is lots of work. For now, we agree to add TogetherAI and mixedbread.ai into MLflow.

You can refer to this PR: https://github.com/mlflow/mlflow/pull/11195 and https://github.com/mlflow/mlflow/pull/11020 for adding new provider

**Comment by FotiosBistas** [comment]
Hey @WeichenXu123 I will try to add the following providers in this order: 

- [ ] TogetherAI
- [ ] Mixedbread
- [ ] Anyscale

Shall I open a seperate PR for each one? I believe this would be more appropriate.

**Comment by BenWilson2** [comment]
@FotiosBistas yes, a separate PR for each (including docs updates and examples for each) would be preferable due to the size of each PR. Thanks!

**Comment by FotiosBistas** [comment]
@WeichenXu123 @BenWilson2 

Just opened the PR: #11557

**Comment by FotiosBistas** [comment]
Hey @WeichenXu123 @BenWilson2,

I added all the available endpoints in the PR for TogetherAI. I think this would be a good time to review it!

**Comment by sir3mat** [comment]
Hi! I have a local model exposed with vllm (with openAi compatible API). Do you think that is possible to allow users change the base_url in order to perform LLM-as-a-judge evaluation with the local model?
I find that these types of evaluation load the api endpoint in this way: https://github.com/mlflow/mlflow/blob/0b1d995d66a678153e01ed3040f3f4dfc16a0d6b/mlflow/metrics/genai/model_utils.py#L98

I try to load an eval by changing that openai endpoint to my local endpoint and it works.

Maybe this could help people with all local models to run experiments with those models and not depending on openAI. 

Here a snippet from my code
```
from llama_index.llms.openai_like import OpenAILike
from typing import Sequence
from llama_index.core.base.llms.types import (
    ChatMessage,
    MessageRole
)

temperature = 0
max_tokens = 1000
client = OpenAILike(
    model=model["name"],
    api_base=model["url"],
    api_key="api-key",
    temperature=temperature,
    max_tokens=max_tokens,
    timeout=180,
)


def openai_like_qa(inputs):
    answers = []
    system_prompt = "Please answer the following question in formal language."
    
    for index, row in inputs.iterrows():
        messages: Sequence[ChatMessage]= [
            ChatMessage(role=MessageRole.SYSTEM, content=system_prompt),
            ChatMessage(role=MessageRole.USER, content=row["inputs"]),
        ]
        completion = client.chat(
            messages,
            model=model_name
        )
        answers.append(completion.message.content)

    return answers
    
with mlflow.start_run(run_name=f"log_demo_MlFlow") as run:
  results = mlflow.evaluate(
          model=openai_like_qa,
          data=eval_data,
          targets="ground_truth",
          predictions="outputs",
          extra_metrics=[mlflow.metrics.genai.answer_similarity(
                  model="openai:/mistralai/Mistral-7B-Instruct-v0.3",
                  )],
          evaluators="default",
      )
      print(f"See aggregated evaluation results below: \n{results.metrics}")
  
      eval_table = results.tables["eval_results_table"]
      print(f"See evaluation table below: \n{eval_table}")
```

and then i change REQUEST_URL_CHAT to my "http://localendpoint:port/v1/chat/completion" inside [mlflow/utils/openai_utils.py (Line 31)](https://github.com/mlflow/mlflow/blob/0b1d995d66a678153e01ed3040f3f4dfc16a0d6b/mlflow/utils/openai_utils.py#L31)
